keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,902 @@
+   if there are many identical physical devices that must be represented as
+   `CounterSet`s. `CounterSetMixins` can not be referenced directly by devices.
+
+1. The `Device` field is a list of named `DeviceMixin`s. These define
+   attributes and capacities that can be used to extend what is defined
+   explicitly in `Device`. `DeviceMixin`s cannot be allocated directly, but can
+   only be referenced by devices.
+
+1. The `DeviceCounterConsumption` field defines a list of named
+   `DeviceCounterConsumptionMixin`s. These define counters that can be
+   used to extend the counter consumption defined explicitly in `DeviceCounterConsumption`
+   in the `ConsumesCounters` list on `Device`. The `CounterSet` from
+   which the counters will be consumed is not specified in the
+   `DeviceCounterConsumptionMixin`, but rather provided when the mixin is
+   referenced from the device.
+
+The mixins are referenced using the same pattern in all three places. The field
+is named `Includes` and will contain a list of references to the mixins. Rather
+than representing the references as a list of strings, there is a list of type
+`<SomeType>MixinRef` which has a `Name` field. This enables adding additional
+values to the reference in the future.
+
+1. The `Includes` field on `CounterSet` is a list of `CounterSetMixinRef`s. This
+   can reference mixins defined in the `CounterSet` field on the `ResourceSliceMixins`.
+
+1. The `Includes` field on `Device` is a list of of `DeviceMixinRef`s. This can
+   reference mixins defined in the `Device` field on the `ResourceSliceMixins`.
+
+1. The `Includes` field on `DeviceCounterConsumption` is a list of 
+   `DeviceCounterConsumptionMixinRef`s. This can reference mixins defined
+   in the `DeviceCounterConsumption` field on the `ResourceSliceMixins`.
+
+With these changes, attributes, capacity, and counters that are shared across
+devices or counter sets can be split out into mixins, thereby reducing
+duplication and reducing the size of the ResourceSlice object.
+
+### Risks and Mitigations
+
+This change doesn't really affect the functionality of DRA, it just
+provides a more compact way to define devices in ResourceSlices. But it
+still have some consequences worth pointing out here.
+
+#### ResourceSlice resources will be harder to understand
+
+The biggest challenge with this change is that it adds a level of
+indirection for the `Device` and `CounterSet` definitions, meaning
+that it gets harder to understand the ResourceSlice objects.
+
+We have discussed adding a kubectl command or a plugin that will allow
+users to see the fully flattened versions of a ResourceSlice. But this
+is not in scope for alpha.
+
+#### More attributes, capacities and counters might worsen worst-case scheduling
+
+With mixins, DRA driver authors can choose to create more complex devices,
+which might lead to worse scheduling performance. But it is up to DRA driver
```

Comment:
I might replace "But it is up to DRA driver authors to do this, so it is not something that can be triggered by normal users." with something like:

"This will not negatively effect existing scheduling performance of existing ResourceSlice definitions, but DRA driver authors taking advantage of mixins should be made aware of possible performance effects due to this increased referential complexity. Furthermore, this reinforces the criticality of ensuring that DRA primitives are optimized for maximum performance."

(That's my takeaway, at least.)

Reply:
I update the KEP to include your suggestion with some small edits.